File: skellig-test-step-processing-jdbc/src/main/java/org/skellig/teststep/processor/jdbc/JdbcTestStepProcessor.java
package org.skellig.teststep.processor.jdbc;

import com.typesafe.config.Config;
import org.skellig.teststep.processing.converter.TestStepResultConverter;
import org.skellig.teststep.processing.processor.TestStepProcessor;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;
import org.skellig.teststep.processor.db.DatabaseTestStepProcessor;
import org.skellig.teststep.processor.db.model.DatabaseTestStep;
import org.skellig.teststep.processor.jdbc.model.JdbcDetails;

import java.util.Map;

public class JdbcTestStepProcessor extends DatabaseTestStepProcessor<JdbcRequestExecutor> {

    protected JdbcTestStepProcessor(Map<String, JdbcRequestExecutor> dbServers,
                                    TestScenarioState testScenarioState,
                                    TestStepResultValidator validator,
                                    TestStepResultConverter testStepResultConverter) {
        super(dbServers, testScenarioState, validator, testStepResultConverter);
    }

    public static class Builder extends DatabaseTestStepProcessor.Builder<JdbcDetails, JdbcRequestExecutor> {

        private JdbcDetailsConfigReader jdbcDetailsConfigReader;

        public Builder() {
            jdbcDetailsConfigReader = new JdbcDetailsConfigReader();
        }

        @Override
        public DatabaseTestStepProcessor.Builder<JdbcDetails, JdbcRequestExecutor> withDbServers(Config config) {
            jdbcDetailsConfigReader.read(config).forEach(this::withDbServer);
            return this;
        }

        @Override
        protected JdbcRequestExecutor createRequestExecutor(JdbcDetails databaseDetails) {
            return new JdbcRequestExecutor(databaseDetails);
        }

        @Override
        public TestStepProcessor<DatabaseTestStep> build() {
            return new JdbcTestStepProcessor(dbServers, testScenarioState, validator, testStepResultConverter);
        }
    }
}

File: skellig-test-step-processing-jdbc/src/main/kotlin/org/skellig/teststep/processor/jdbc/JdbcTestStepProcessor.kt
package org.skellig.teststep.processor.jdbc

import com.typesafe.config.Config
import org.skellig.teststep.processing.processor.TestStepProcessor
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processor.db.DatabaseTestStepProcessor
import org.skellig.teststep.processor.jdbc.model.JdbcDetails
import org.skellig.teststep.processor.jdbc.model.JdbcTestStep

/**
 * JdbcTestStepProcessor is a class that processes JDBC test steps for executing database requests.
 *
 * @property dbServers a map of database servers with their corresponding [JdbcRequestExecutor], where key is a server name from [servers][JdbcTestStep.servers]
 * @property testScenarioState the [TestScenarioState] object that holds the state of the test scenario.
 */
open class JdbcTestStepProcessor protected constructor(dbServers: Map<String, JdbcRequestExecutor>,
                                                       testScenarioState: TestScenarioState)
    : DatabaseTestStepProcessor<JdbcRequestExecutor, JdbcTestStep>(dbServers, testScenarioState) {

    override fun getTestStepClass(): Class<*> = JdbcTestStep::class.java

    class Builder : DatabaseTestStepProcessor.Builder<JdbcDetails, JdbcTestStep, JdbcRequestExecutor>() {

        private val jdbcDetailsConfigReader: JdbcDetailsConfigReader = JdbcDetailsConfigReader()

        override fun withDbServers(config: Config) = apply {
            jdbcDetailsConfigReader.read(config).forEach { databaseDetails: JdbcDetails -> withDbServer(databaseDetails) }
        }

        override fun createRequestExecutor(databaseDetails: JdbcDetails): JdbcRequestExecutor {
            return JdbcRequestExecutor(databaseDetails)
        }

        override fun build(): TestStepProcessor<JdbcTestStep> {
            return JdbcTestStepProcessor(dbServers, testScenarioState!!)
        }
    }
}